### PR TITLE
feat: HA support for event sources and sensors

### PR DESCRIPTION
**Pull request by whynowy**
Signed-off-by: Derek Wang <whynowy@gmail.com>

1. `Active-Passive` HA for some of event sources (the rest ones have been able to do `Active-Active`);
2. `Active-Passive` HA for sensors.
3. `spec.replicas` for Sensor spec.
4. `spec.replicas` for EventSource spec (deprecate `spec.replica`). 

Mitigate #985

Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-events/blob/master/USERS.md).

### CODE REVIEW

File: eventsources/cmd/main.go [hunk 1]
```diff
@@ -58,8 +59,10 @@ func main() {
 	ctx := logging.WithLogger(signals.SetupSignalHandler(), logger)
 	m := metrics.NewMetrics(eventSource.Namespace)
 	go m.Run(ctx, fmt.Sprintf(":%d", common.EventSourceMetricsPort))
+
+	logger.Infow("starting eventsource server", "version", argoevents.GetVersion())
 	adaptor := eventsources.NewEventSourceAdaptor(eventSource, busConfig, ebSubject, hostname, m)
 	if err := adaptor.Start(ctx); err != nil {
-		logger.Desugar().Fatal("failed to start eventsource server", zap.Error(err))
+		logger.Fatalw("failed to start eventsource server", zap.Error(err))
 	}
```
Comment (alexec) [inline eventsources/cmd/main.go:67]: Maybe it is time to have a PR to de-desugar the whole codebase at once?

File: eventsources/eventing.go [hunk 2]
```diff
@@ -263,20 +268,90 @@ func NewEventSourceAdaptor(eventSource *v1alpha1.EventSource, eventBusConfig *ev
...
 	servers := GetEventingServers(e.eventSource, e.metrics)
+	for k := range servers {
+		if _, ok := recreateTypes[k]; ok {
+			isRecreatType = true
+		}
+		// This is based on the presumption that all the events in one
+		// EventSource object use the same type of deployment strategy
+		break
+	}
+	if !isRecreatType || e.eventSource.Spec.GetReplicas() == 1 {
```
Comment (alexec) [inline eventsources/eventing.go:287]: Minor - is workflows we always use leader election.  What happens when you scale from 1 to 2 and the first replica is not already using it? Don’t you get two leaders?

File: eventsources/eventing.go [hunk 3]
```diff
@@ -263,20 +268,90 @@ func NewEventSourceAdaptor(eventSource *v1alpha1.EventSource, eventBusConfig *ev
...
+
+	leaseName := "eventsource-" + e.eventSource.Name
+	lock := &resourcelock.LeaseLock{
+		LeaseMeta: metav1.ObjectMeta{
+			Name:      leaseName,
+			Namespace: e.eventSource.Namespace,
+		},
+		Client: kubeClient.CoordinationV1(),
+		LockConfig: resourcelock.ResourceLockConfig{
+			Identity: e.hostname,
```
Comment (alexec) [inline eventsources/eventing.go:306]: I think we also emit events on leadership change in workflows

File: pkg/apis/sensor/v1alpha1/types.go [hunk 4]
```diff
@@ -111,6 +111,19 @@ type SensorSpec struct {
 	// Circuit is a boolean expression of dependency groups
 	// Deprecated: will be removed in v1.5, use Switch in triggers instead.
 	DeprecatedCircuit string `json:"circuit,omitempty" protobuf:"bytes,7,opt,name=circuit"`
+	// Replicas is the sensor deployment replicas
+	Replicas *int32 `json:"replicas,omitempty" protobuf:"varint,8,opt,name=replicas"`
+}
+
+func (s SensorSpec) GetReplicas() int32 {
+	if s.Replicas == nil {
```
Comment (alexec) [inline pkg/apis/sensor/v1alpha1/types.go:119]: Test?

File: pkg/apis/eventsource/v1alpha1/types.go [hunk 5]
```diff
@@ -109,6 +110,24 @@ type EventSourceSpec struct {
 	Pulsar map[string]PulsarEventSource `json:"pulsar,omitempty" protobuf:"bytes,27,opt,name=pulsar"`
 	// Generic event source
 	Generic map[string]GenericEventSource `json:"generic,omitempty" protobuf:"bytes,28,rep,name=generic"`
+	// Replicas is the event source deployment replicas
+	Replicas *int32 `json:"replicas,omitempty" protobuf:"varint,29,opt,name=replicas"`
+}
+
+func (e EventSourceSpec) GetReplicas() int32 {
+	if e.Replicas == nil && e.DeprecatedReplica == nil {
```
Comment (alexec) [inline pkg/apis/eventsource/v1alpha1/types.go:118]: This is just complicated enough to warrant a unit test.

File: test/util/util.go [hunk 6]
```diff
@@ -265,7 +284,24 @@ func SensorPodLogContains(ctx context.Context, kubeClient kubernetes.Interface,
 		return false, fmt.Errorf("error getting sensor pod name: %w", err)
 	}
 	podName := podList.Items[0].GetName()
-	fmt.Printf("Sensor POD name: %s", podName)
+	if len(podList.Items) > 1 {
```
Comment (alexec) [inline test/util/util.go:287]: I think you can simplify the code by removing this guard and the outcome should be the same?

File: test/util/util.go [hunk 7]
```diff
@@ -252,6 +254,23 @@ func EventSourcePodLogContains(ctx context.Context, kubeClient kubernetes.Interf
 		return false, fmt.Errorf("error getting event source pod name: %w", err)
 	}
 	podName := podList.Items[0].GetName()
+	if len(podList.Items) > 1 {
```
Comment (alexec) [inline test/util/util.go:257]: I think you can simplify the code by removing this guard condition.

**Review by alexec — COMMENTED** [review]
@sarabala1979 should review too? How did you test this?

File: eventsources/eventing.go (hunk 2 above)
Comment (sarabala1979) [reply eventsources/eventing.go:287]: I would also recommend using leader election by default which will avoid the code complexity to deal separately. Workflow enables election leader by default

**Comment by sarabala1979** [comment]
> @sarabala1979 should review too? How did you test this?
I tested locally switching the pods (killing current leader) and verify the logs for leader switching.

File: pkg/apis/sensor/v1alpha1/types.go (hunk 4 above)
Comment (whynowy) [reply pkg/apis/sensor/v1alpha1/types.go:119]: Test cases added.

File: pkg/apis/eventsource/v1alpha1/types.go (hunk 5 above)
Comment (whynowy) [reply pkg/apis/eventsource/v1alpha1/types.go:118]: Test cases added.

File: eventsources/eventing.go (hunk 3 above)
Comment (alexec) [reply eventsources/eventing.go:306]: good point

File: eventsources/eventing.go (hunk 2 above)
Comment (alexec) [reply eventsources/eventing.go:287]: Ah - so users must all install the extra RBAC - a process that does not scale. Good point.

**Review by alexec — APPROVED** [review]

**Pull request merged**